File: Sources/PalmierPro/Timeline/TimelineGeometry.swift
import AppKit

enum TrackDropTarget: Equatable {
    case existingTrack(Int)
    case newTrackAt(Int) // insert new track before this index
}

/// Pure layout math for the timeline. Used by both TimelineView (drawing)
/// and TimelineInputController (hit testing).
struct TimelineGeometry {
    let pixelsPerFrame: Double
    let headerWidth: Double
    let rulerHeight: CGFloat
    let trackCount: Int
    let trackHeights: [CGFloat]
    let bounds: NSRect

    /// Precomputed cumulative Y offsets for each track (avoids O(n) per lookup).
    private let cumulativeY: [CGFloat]

    @MainActor
    init(editor: EditorViewModel, bounds: NSRect, headerWidth: Double = 0) {
        self.pixelsPerFrame = editor.zoomScale
        self.headerWidth = headerWidth
        self.rulerHeight = Layout.rulerHeight
        self.trackCount = editor.timeline.tracks.count
        self.trackHeights = editor.timeline.tracks.map(\.displayHeight)
        self.bounds = bounds

        var cumY: [CGFloat] = []
        cumY.reserveCapacity(trackHeights.count)
        var y = rulerHeight + Layout.dropZoneHeight
        for h in trackHeights {
            cumY.append(y)
            y += h
        }
        self.cumulativeY = cumY
    }

    func trackHeight(at index: Int) -> CGFloat {
        trackHeights.indices.contains(index) ? trackHeights[index] : Layout.trackHeight
    }

    func trackY(at index: Int) -> CGFloat {
        cumulativeY.indices.contains(index) ? cumulativeY[index] : rulerHeight
    }

    func clipRect(for clip: Clip, trackIndex: Int) -> NSRect {
        clipRect(for: clip, atY: Double(trackY(at: trackIndex)), height: Double(trackHeight(at: trackIndex)))
    }

    /// Clip rect at an arbitrary Y position (used for ghost clips at insertion lines).
    func clipRect(for clip: Clip, atY y: Double, height h: Double) -> NSRect {
        NSRect(
            x: headerWidth + Double(clip.startFrame) * pixelsPerFrame,
            y: y + 2,
            width: Double(clip.durationFrames) * pixelsPerFrame,
            height: h - 4
        )
    }

    func frameAt(x: Double) -> Int {
        max(0, Int((x - headerWidth) / pixelsPerFrame))
    }

    func trackAt(y: Double) -> Int {
        for i in cumulativeY.indices {
            if y < Double(cumulativeY[i]) + Double(trackHeights[i]) { return i }
        }
        return max(0, trackCount - 1)
    }

    func dropTargetAt(y: Double) -> TrackDropTarget {
        guard trackCount > 0 else { return .newTrackAt(0) }

        // Top drop zone
        if y < Double(cumulativeY[0]) {
            return .newTrackAt(0)
        }

        // Check between-track boundaries
        let threshold = Double(Layout.insertThreshold)
        for i in 0..<(trackCount - 1) {
            let bottomOfTrack = Double(cumulativeY[i]) + Double(trackHeights[i])
            let topOfNext = Double(cumulativeY[i + 1])
            // The boundary region: threshold above the gap to threshold below
            if y >= bottomOfTrack - threshold && y <= topOfNext + threshold {
                return .newTrackAt(i + 1)
            }
        }

        // Bottom drop zone: past the last track
        let lastTrackBottom = Double(cumulativeY[trackCount - 1]) + Double(trackHeights[trackCount - 1])
        if y >= lastTrackBottom {
            return .newTrackAt(trackCount)
        }

        // On an existing track
        for i in cumulativeY.indices {
            if y < Double(cumulativeY[i]) + Double(trackHeights[i]) { return .existingTrack(i) }
        }
        return .existingTrack(max(0, trackCount - 1))
    }

    func insertionLineY(for target: TrackDropTarget) -> CGFloat? {
        switch target {
        case .existingTrack:
            return nil
        case .newTrackAt(let index):
            if trackCount == 0 {
                return rulerHeight + Layout.dropZoneHeight
            } else if index == 0 {
                return cumulativeY[0]
            } else if index >= trackCount {
                return cumulativeY[trackCount - 1] + trackHeights[trackCount - 1]
            } else {
                return cumulativeY[index]
            }
        }
    }

    /// Y position where a ghost clip should render for a new-track drop.
    func ghostY(for target: TrackDropTarget, height: CGFloat = Layout.trackHeight) -> CGFloat? {
        guard case .newTrackAt(let index) = target,
              let lineY = insertionLineY(for: target) else { return nil }
        return index < trackCount ? lineY - height : lineY
    }

    func xForFrame(_ frame: Int) -> Double {
        headerWidth + Double(frame) * pixelsPerFrame
    }

    /// Centerline X of an audio fade handle
    static func audioFadeHandleX(in clipRect: NSRect, fadeFrames: Int, edge: FadeEdge, pxPerFrame: CGFloat) -> CGFloat {
        let fadePx = CGFloat(fadeFrames) * pxPerFrame
        let inset = ClipRenderer.fadeHandleEdgeInset
        let knee = max(inset, min(max(inset, clipRect.width - inset), fadePx))
        return edge == .left ? clipRect.minX + knee : clipRect.maxX - knee
    }

    /// Hit rect for an audio fade handle
    func audioFadeHandleRect(in clipRect: NSRect, fadeFrames: Int, edge: FadeEdge) -> NSRect {
        let cx = Self.audioFadeHandleX(in: clipRect, fadeFrames: fadeFrames, edge: edge, pxPerFrame: pixelsPerFrame)
        let cy = clipRect.minY + ClipRenderer.labelBarHeight
        let half = ClipRenderer.fadeHandleHitSize / 2
        return NSRect(x: cx - half, y: cy - half, width: half * 2, height: half * 2)
    }
}

File: platforms/windows/src/PalmierPro.App/Editing/TimelineGeometry.cs
using PalmierPro.Core.Models;

namespace PalmierPro.App.Editing;

/// Where a track-header drop lands: on an existing track, or inserting a new one before `Index`.
/// Ported from Timeline/TimelineGeometry.swift's `TrackDropTarget` enum.
public abstract record TrackDropTarget
{
    public sealed record ExistingTrack(int Index) : TrackDropTarget;

    public sealed record NewTrackAt(int Index) : TrackDropTarget;
}

/// Axis-aligned rect in timeline-canvas pixels. A local, dependency-free stand-in for AppKit's
/// `NSRect` (same field shape — origin + size, Y-down) so this file stays free of any WinUI/WinRT
/// type and is safe to unit test under plain `dotnet test`.
public readonly record struct EditorRect(double X, double Y, double Width, double Height);

/// Pure layout math for the timeline, ported near-verbatim from Timeline/TimelineGeometry.swift.
/// Used by both the timeline drawing surface and hit-testing during drag — neither exists yet on
/// Windows (Stage D/E), but the math is settled now so both build against it unchanged.
public sealed class TimelineGeometry
{
    /// Layout constants mirrored from Utilities/Constants.swift's `Layout` enum — the subset this
    /// class needs. Kept local rather than added to `PalmierPro.Core` (whose `LayoutDefaults`
    /// deliberately excludes UI-layout constants; see that file's doc comment).
    public static class Layout
    {
        public const double RulerHeight = 24;
        public const double DropZoneHeight = 60;
        public const double TrackHeight = 50;
        public const double InsertThreshold = 10;
        /// Mirrors Mac's `Layout.trackHeaderWidth` — used by the Win2D timeline view (Stage C) for
        /// the fixed-left track-header column.
        public const double HeaderWidth = 100;
    }

    public double PixelsPerFrame { get; }
    public double HeaderWidth { get; }
    public double RulerHeight { get; }
    public int TrackCount { get; }
    public IReadOnlyList<double> TrackHeights { get; }

    /// Precomputed cumulative Y offsets for each track (avoids O(n) per lookup).
    private readonly List<double> _cumulativeY;

    public TimelineGeometry(double pixelsPerFrame, IReadOnlyList<double> trackHeights, double headerWidth = 0)
    {
        PixelsPerFrame = pixelsPerFrame;
        HeaderWidth = headerWidth;
        RulerHeight = Layout.RulerHeight;
        TrackCount = trackHeights.Count;
        TrackHeights = trackHeights;

        _cumulativeY = new List<double>(trackHeights.Count);
        var y = RulerHeight + Layout.DropZoneHeight;
        foreach (var h in trackHeights)
        {
            _cumulativeY.Add(y);
            y += h;
        }
    }

    public double TrackHeight(int index) =>
        index >= 0 && index < TrackHeights.Count ? TrackHeights[index] : Layout.TrackHeight;

    public double TrackY(int index) =>
        index >= 0 && index < _cumulativeY.Count ? _cumulativeY[index] : RulerHeight;

    public EditorRect ClipRect(Clip clip, int trackIndex) =>
        ClipRect(clip, TrackY(trackIndex), TrackHeight(trackIndex));

    /// Clip rect at an arbitrary Y position (used for ghost clips at insertion lines).
    public EditorRect ClipRect(Clip clip, double y, double height) => new(
        HeaderWidth + clip.StartFrame * PixelsPerFrame,
        y + 2,
        clip.DurationFrames * PixelsPerFrame,
        height - 4);

    public int FrameAt(double x) => Math.Max(0, (int)((x - HeaderWidth) / PixelsPerFrame));

    public int TrackAt(double y)
    {
        for (var i = 0; i < _cumulativeY.Count; i++)
        {
            if (y < _cumulativeY[i] + TrackHeights[i])
            {
                return i;
            }
        }
        return Math.Max(0, TrackCount - 1);
    }

    public TrackDropTarget DropTargetAt(double y)
    {
        if (TrackCount == 0)
        {
            return new TrackDropTarget.NewTrackAt(0);
        }

        // Top drop zone.
        if (y < _cumulativeY[0])
        {
            return new TrackDropTarget.NewTrackAt(0);
        }

        // Check between-track boundaries.
        const double threshold = Layout.InsertThreshold;
        for (var i = 0; i < TrackCount - 1; i++)
        {
            var bottomOfTrack = _cumulativeY[i] + TrackHeights[i];
            var topOfNext = _cumulativeY[i + 1];
            // The boundary region: threshold above the gap to threshold below.
            if (y >= bottomOfTrack - threshold && y <= topOfNext + threshold)
            {
                return new TrackDropTarget.NewTrackAt(i + 1);
            }
        }

        // Bottom drop zone: past the last track.
        var lastTrackBottom = _cumulativeY[TrackCount - 1] + TrackHeights[TrackCount - 1];
        if (y >= lastTrackBottom)
        {
            return new TrackDropTarget.NewTrackAt(TrackCount);
        }

        // On an existing track.
        for (var i = 0; i < _cumulativeY.Count; i++)
        {
            if (y < _cumulativeY[i] + TrackHeights[i])
            {
                return new TrackDropTarget.ExistingTrack(i);
            }
        }
        return new TrackDropTarget.ExistingTrack(Math.Max(0, TrackCount - 1));
    }

    public double? InsertionLineY(TrackDropTarget target)
    {
        if (target is not TrackDropTarget.NewTrackAt(var index))
        {
            return null;
        }
        if (TrackCount == 0)
        {
            return RulerHeight + Layout.DropZoneHeight;
        }
        if (index == 0)
        {
            return _cumulativeY[0];
        }
        if (index >= TrackCount)
        {
            return _cumulativeY[TrackCount - 1] + TrackHeights[TrackCount - 1];
        }
        return _cumulativeY[index];
    }

    /// Y position where a ghost clip should render for a new-track drop.
    public double? GhostY(TrackDropTarget target, double height = Layout.TrackHeight)
    {
        if (target is not TrackDropTarget.NewTrackAt(var index) || InsertionLineY(target) is not { } lineY)
        {
            return null;
        }
        return index < TrackCount ? lineY - height : lineY;
    }

    public double XForFrame(int frame) => HeaderWidth + frame * PixelsPerFrame;
}
